spatial/r3/vector.go

```diff
@@ -95,6 +95,19 @@ func Divergence(p, step Vec, field func(Vec) Vec) float64 {
 	return 0.5 * (divx + divy + divz)
 }
 
+// Gradient returns the gradient of the scalar field at the point p,
+// approximated using finite differences with the given step sizes.
+func Gradient(p, step Vec, field func(Vec) float64) Vec {
+	dx := Vec{X: step.X}
+	dy := Vec{Y: step.Y}
+	dz := Vec{Z: step.Z}
+	return Vec{
+		X: field(Add(p, dx)) - field(Sub(p, dx)),
```

Comment:
Surely there must be a non-obvious more performant solution. This was the simplest one I could conjure. I'll see what I can find on the web

Reply:
You don't have to change anything, it was really just a question. I'm referring to forming the evaluation points with `Add` and `Sub` knowing that 2 out of 3 components are zero. I wouldn't search on the web, I would write a benchmark and compare this implementation with one that for example builds the evaluation points explicitly without Add/Sub.

Reply:
```go
func BenchmarkGradient1(b *testing.B) {
	step := Vec{.1, .1, .1}
	rnd := rand.New(rand.NewSource(1))
	for i := 0; i < b.N; i++ {
		p := randomVec(rnd)
		got := Gradient(p, step, scalarFields[0].field)
		_ = got
	}
}

// Gradient returns the gradient of the scalar field at the point p,
// approximated using finite differences with the given step sizes.
func Gradient2(p, step Vec, field func(Vec) float64) Vec {
	ppx := Vec{p.X + step.X, p.Y, p.Z}
	pmx := Vec{p.X - step.X, p.Y, p.Z}
	ppy := Vec{p.X, p.Y + step.Y, p.Z}
	pmy := Vec{p.X, p.Y - step.Y, p.Z}
	pmz := Vec{p.X, p.Y, p.Z + step.Z}
	ppz := Vec{p.X, p.Y, p.Z - step.Z}
	return Vec{
		X: field(ppx) - field(pmx),
		Y: field(ppy) - field(pmy),
		Z: field(ppz) - field(pmz),
	}
}

func BenchmarkGradient2(b *testing.B) {
	step := Vec{.1, .1, .1}
	rnd := rand.New(rand.NewSource(1))
	for i := 0; i < b.N; i++ {
		p := randomVec(rnd)
		got := Gradient2(p, step, scalarFields[0].field)
		_ = got
	}
}
```
Results:
```
$ go test ./spatial/r3/ -bench=.
goos: linux
goarch: amd64
pkg: gonum.org/v1/gonum/spatial/r3
cpu: 12th Gen Intel(R) Core(TM) i5-12400F
BenchmarkQuat-12                20230939               111.9 ns/op
BenchmarkGradient1-12           62112535                19.10 ns/op
BenchmarkGradient2-12           53097438                19.90 ns/op
PASS
ok      gonum.org/v1/gonum/spatial/r3   5.251s
```

Reply:
When benchmarking it's good to specify `-count` and use `benchstat` https://pkg.go.dev/golang.org/x/perf/cmd/benchstat.

I did:
```
» go test -v -run XXX -bench Grad -count 10 > old.txt
» # Change Gradient to the "new" implementation.
» go test -v -run XXX -bench Grad -count 10 > new.txt
» benchstat old.txt new.txt
name        old time/op  new time/op  delta
Gradient-4  36.4ns ± 5%  35.9ns ± 2%   ~     (p=0.071 n=9+7)
```
which shows there's no significant difference between the two versions. And even if there were some, it would go away as soon as the evaluation of `field` were more expensive.